File: tools/ao-export/export-bmp-to-png.mjs
/**
 * BMP AO → PNG (fondo negro = transparente)
 *
 * npm install
 * npm run bmp2png -- --bmp-dir "C:\...\RecursosAO\Recursos\Graficos\bmp" --out-dir "C:\...\AOWEB\public\assets\ao\png"
 */

import fs from "fs";
import path from "path";
import bmp from "bmp-js";
import { PNG } from "pngjs";

const TOLERANCE = 8;

function parseArgs() {
  const args = process.argv.slice(2);
  const opts = { limit: 0 };
  const positional = [];

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--bmp-dir") opts.bmpDir = args[++i];
    else if (args[i] === "--out-dir") opts.outDir = args[++i];
    else if (args[i] === "--limit") opts.limit = Number(args[++i]);
    else if (!args[i].startsWith("--")) positional.push(args[i]);
  }

  if (!opts.bmpDir && positional[0]) opts.bmpDir = positional[0];
  if (!opts.outDir && positional[1]) opts.outDir = positional[1];

  return opts;
}

function isTransparent(r, g, b) {
  return (
    Math.abs(r) <= TOLERANCE &&
    Math.abs(g) <= TOLERANCE &&
    Math.abs(b) <= TOLERANCE
  );
}

function bmpToPng(srcPath, dstPath) {
  const buf = fs.readFileSync(srcPath);
  const { data, width, height } = bmp.decode(buf);
  const png = new PNG({ width, height });

  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const i = (width * y + x) * 4;
      const r = data[i];
      const g = data[i + 1];
      const b = data[i + 2];
      const idx = (width * y + x) << 2;
      png.data[idx] = r;
      png.data[idx + 1] = g;
      png.data[idx + 2] = b;
      png.data[idx + 3] = isTransparent(r, g, b) ? 0 : 255;
    }
  }

  fs.mkdirSync(path.dirname(dstPath), { recursive: true });
  fs.writeFileSync(dstPath, PNG.sync.write(png));
}

const { bmpDir, outDir, limit } = parseArgs();

if (!bmpDir || !outDir) {
  console.log("Uso:");
  console.log(
    '  npm run bmp2png -- --bmp-dir "RUTA\\Graficos\\bmp" --out-dir "RUTA\\salida\\png"'
  );
  process.exit(1);
}

if (!fs.existsSync(bmpDir)) {
  console.error(`ERROR: No existe la carpeta:\n  ${bmpDir}`);
  console.error("\nCopiá los .bmp desde Argentum Online (Steam) — ver README.md");
  process.exit(1);
}

const files = fs
  .readdirSync(bmpDir)
  .filter((f) => f.toLowerCase().endsWith(".bmp"))
  .sort((a, b) => a.localeCompare(b, undefined, { numeric: true }));

if (files.length === 0) {
  console.error(`ERROR: No hay .bmp en:\n  ${bmpDir}`);
  process.exit(1);
}

const toProcess = limit > 0 ? files.slice(0, limit) : files;
console.log(`Convirtiendo ${toProcess.length} archivos...`);
console.log(`  Origen: ${bmpDir}`);
console.log(`  Destino: ${outDir}`);

for (let i = 0; i < toProcess.length; i++) {
  const file = toProcess[i];
  const stem = path.parse(file).name;
  bmpToPng(path.join(bmpDir, file), path.join(outDir, `${stem}.png`));
  if ((i + 1) % 200 === 0 || i + 1 === toProcess.length) {
    console.log(`  ${i + 1}/${toProcess.length}`);
  }
}

console.log("Listo.");

File: tools/ao-export/export_bmp_to_png.py
"""
Convierte BMP de Argentum Online a PNG con fondo negro transparente.
Uso:
  python export_bmp_to_png.py --bmp-dir RUTA_A_Graficos/bmp --out-dir RUTA_SALIDA
"""

from __future__ import annotations

import argparse
from pathlib import Path

from PIL import Image

# En el cliente AO el negro puro suele ser transparente.
TRANSPARENT_RGB = (0, 0, 0)
TOLERANCE = 8


def bmp_to_png(src: Path, dst: Path) -> None:
    img = Image.open(src).convert("RGBA")
    pixels = img.load()
    w, h = img.size

    for y in range(h):
        for x in range(w):
            r, g, b, a = pixels[x, y]
            if (
                abs(r - TRANSPARENT_RGB[0]) <= TOLERANCE
                and abs(g - TRANSPARENT_RGB[1]) <= TOLERANCE
                and abs(b - TRANSPARENT_RGB[2]) <= TOLERANCE
            ):
                pixels[x, y] = (0, 0, 0, 0)

    dst.parent.mkdir(parents=True, exist_ok=True)
    img.save(dst, "PNG")


def main() -> None:
    parser = argparse.ArgumentParser(description="BMP AO → PNG con transparencia")
    parser.add_argument(
        "--bmp-dir",
        required=True,
        help="Carpeta con archivos .bmp (ej. RecursosAO/Recursos/Graficos/bmp)",
    )
    parser.add_argument(
        "--out-dir",
        required=True,
        help="Carpeta de salida para los .png",
    )
    parser.add_argument(
        "--limit",
        type=int,
        default=0,
        help="Máximo de archivos a convertir (0 = todos). Útil para probar.",
    )
    args = parser.parse_args()

    bmp_dir = Path(args.bmp_dir)
    out_dir = Path(args.out_dir)

    if not bmp_dir.is_dir():
        print(f"ERROR: No existe la carpeta BMP:\n  {bmp_dir}")
        print("\nCopiá los .bmp desde la instalación de AO (Steam) — ver README.md")
        return

    files = sorted(bmp_dir.glob("*.bmp")) + sorted(bmp_dir.glob("*.BMP"))
    if not files:
        print(f"ERROR: No hay archivos .bmp en:\n  {bmp_dir}")
        return

    if args.limit > 0:
        files = files[: args.limit]

    total = len(files)
    print(f"Convirtiendo {total} archivos...")
    print(f"  Origen: {bmp_dir}")
    print(f"  Destino: {out_dir}")

    for i, bmp_path in enumerate(files, 1):
        png_path = out_dir / f"{bmp_path.stem}.png"
        bmp_to_png(bmp_path, png_path)
        if i % 200 == 0 or i == total:
            print(f"  {i}/{total}")

    print("Listo.")


if __name__ == "__main__":
    main()
